### ISSUE TITLE
Netlify CLI deploy command error when deploying golang binary : Invalid AWS Lambda parameters. Reserved keys used in this request

### ISSUE THREAD

**Issue by wahyudibo** [netlify/cli#1147]
Hi, i want to try netlify to host my golang project. I clone the hello world project from here https://github.com/netlify/aws-lambda-go-example and have installed netlify via `npm install netlify-cli -g`. The `netlify build` command had run successfully and the binary is created under `functions` folder. The problem happens when i run `netlify deploy --prod` and it gives me this error (took with DEBUG on): 

```
✔ CDN requesting 0 files and 1 functions
▐|\____________▌ (0/1) Uploading hello-lambda...Warning: JSONHTTPError: undefined 500
    at Object.warn (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/node_modules/@oclif/errors/lib/index.js:49:15)
    at DeployCommand.warn (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/node_modules/@oclif/command/lib/command.js:57:16)
    at DeployCommand.run (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/src/commands/deploy.js:216:16)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
Warning:
{
  "name": "JSONHTTPError",
  "status": 500,
  "json": {
    "error": "Invalid AWS Lambda parameters. Reserved keys used in this request"
  }
}

    at Object.warn (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/node_modules/@oclif/errors/lib/index.js:49:15)
    at DeployCommand.warn (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/node_modules/@oclif/command/lib/command.js:57:16)
    at DeployCommand.run (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/src/commands/deploy.js:217:16)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
▐_|\___________▌ (0/1) Uploading hello-lambda...JSONHTTPError: Internal Server Error
    at parseResponse (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/node_modules/netlify/src/methods/response.js:12:11)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async callMethod (/home/wahyudibo/.nvm/versions/node/v12.18.0/lib/node_modules/netlify-cli/node_modules/netlify/src/methods/index.js:36:26)
```

i already tried googling with `Invalid AWS Lambda parameters. Reserved keys used in this request` keyword and stumbled upon this thread : https://community.netlify.com/t/build-fails-invalid-aws-lambda-parameters/4235/3 and the case is almost similar. The difference is the solution mentioned that 

```
the restriction of the maximum size of environment variables concatenated all together (e.g. VAR1=val1,VAR2=val2) should not be bigger than 4,096 bytes.
``` 

but it confuses me more since i have no environment variables to begin with since this is a hello world project. I also take a look at the reserved environment variable at https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime but found that my environment variables is not conflicted with the one mentioned in AWS documentation.

My question is: If `netlify deploy` takes into account my local environment variable size so it violates the 4096 rules, what is the workaround to deploy my functions and site to netlify using netlify cli ? i know there are ways such using trigger from git commits but i want to stick with `netlify deploy` command if it's possible. I also read in documentation that it should be run without problem : https://cli.netlify.com/commands/deploy
 
Thanks in advance!

PS:
i'm using go mod and here is my Makefile

```
build:
	GO111MODULE="on"
	mkdir -p functions
	go get ./...
	go build -o functions/hello-lambda ./...
```

**Comment by azarai** [comment]
Did run into the same problem today. Fresh project following the example, also using go mod, no env variables set for the site.

Also, ran `netlify deploy` in a shell without my systems env variables (from [stackexchange](https://unix.stackexchange.com/questions/48994/how-to-run-a-program-in-a-clean-environment-in-bash)) with `env -i bash --noprofile --norc` and manually adding the PATH so I could run node and netlify. Same result.

Any fixes or workarounds for that?

**Comment by erezrokah** [comment]
Hi @azarai, do you mind providing a step by step reproduction possibly with a public repo?
I wasn't able to reproduce this using https://github.com/netlify/aws-lambda-go-example

**Comment by azarai** [comment]
> Hi @azarai, do you mind providing a step by step reproduction possibly with a public repo?
> I wasn't able to reproduce this using https://github.com/netlify/aws-lambda-go-example

Sure, I used https://github.com/azarai/go-netlify-test

With netlify deploy I ran into the above problem. Once I used the Git deploy way in Netlify it worked without any changes.

**Comment by erezrokah** [comment]
Hi @azarai, I'm getting 404 on that link so the repo is probably private.

**Comment by azarai** [comment]
> Hi @azarai, I'm getting 404 on that link so the repo is probably private.

Ah, sorry. Made that public now.

**Comment by jmcvetta** [comment]
I'm seeing the same error, both on my local machine and when running from Github Actions (which I think has a smaller set of environment variables than a typical shell).

**Comment by jmcvetta** [comment]
I found a solution.  My `netlify.toml` file specified an environment variable named `BASE_URL`.  Apparently that is a reserved key for Lambda.  

Updating my code and Netlify config to use an env var name like `MY_APP_BASE_URL` solved the problem.

**Comment by romain130492** [comment]
any fix ?...

**Comment by lsnow99** [comment]
Hey I'm running into the same issue- any known workarounds?

**Comment by patrickdappollonio** [comment]
It seems the workaround is letting Netlify build the function for you, instead of you doing it locally. There's something wrong between the build-and-pack part when you do it yourself. I opened a ticket with Netlify and they said the issue comes from "how the zip is being made" -- but not sure what zip are they talking about. 

If anyone has found out a better workaround please share!

**Comment by lsnow99** [comment]
Thanks @patrickdappollonio . Unfortunately my website has other build steps that make it ideal to do in a GitHub CI job as opposed to taking advantage of the netlify build feature. If the ticket is public please do share the link to it. I have an open post on the support forums currently: https://answers.netlify.com/t/invalid-aws-lambda-parameters-reserved-keys-used-in-this-request/47233/8

**Comment by patrickdappollonio** [comment]
@lsnow99 I'm also looking for a solution since I'm working the same way as you, where my pipeline is in Github and I would rather keep it there for unit tests and the likes.  The ticket isn't public but if you're a paying Netlify customer you can e-mail them. 

The same issue is actually outlined in another Github repo from Netlify: https://github.com/netlify/go-functions-example/issues/2 -- that repository also does not work (unless you have them build it for you, which defeats CI/CD solutions already in place)

**Comment by cbmdfc** [comment]
judging by the fact this ticket has been opened for well over a year, I doubt @netlify wants to fix it, they would rather see you pay them the build time

**Comment by patrickdappollonio** [comment]
A quick update for anyone waiting: the ticket I opened as a paying customer was considered "complete" because I commented in this thread and, as such, "I'll be notified when the issue is resolved".

Judging that both this Github issue has been longstanding, and the other github repository about Go on Netlify is quite old too, it's safe to assume this is not in their priority list. I'm moving on to Vercel and/or Github pages depending on the kind of service I need to have.

**Comment by eduardoboucas** [comment]
Hi all! First of all, apologies for the really slow response on this issue.

We've been working on a number of improvements to our Go functions offering, namely the new local development experience using the CLI and the ability to build the functions from source. Note that this doesn't mean we're forcing people to use our build system, because you can build the functions wherever you like by running `netlify deploy --build`. Have you tried using this command?

We still want to address the specific use case of folks who prefer to build the binaries themselves, and I'm sorry that this has been a frustrating experience. We'll look into this and provide an update by the end of this week.

Thanks for your patience.

**Comment by patrickdappollonio** [comment]
I tried your suggestion using `netlify deploy --build` by cloning the repository: https://github.com/netlify/go-functions-example

I managed to get it to build locally, however, the outcome is the same:

```
$ netlify deploy --build

────────────────────────────────────────────────────────────────
  Netlify Build
────────────────────────────────────────────────────────────────

❯ Version
  @netlify/build 26.1.7

❯ Flags
  {}

❯ Current directory
  /home/go/temp-f7355ed7/go-functions-example

❯ Config file
  /home/go/temp-f7355ed7/go-functions-example/netlify.toml

❯ Context
  production

────────────────────────────────────────────────────────────────
  1. build.command from netlify.toml
────────────────────────────────────────────────────────────────

$ make build
mkdir -p functions
go get ./...
go get: added github.com/aws/aws-lambda-go v1.28.0
go build -o functions/hello-lambda ./...

(build.command completed in 1.6s)

────────────────────────────────────────────────────────────────
  2. Functions bundling
────────────────────────────────────────────────────────────────

Packaging Functions from functions directory:
 - hello-lambda


(Functions bundling completed in 49ms)

────────────────────────────────────────────────────────────────
  Netlify Build Complete
────────────────────────────────────────────────────────────────

(Netlify Build completed in 1.6s)
Deploy path:        /home/go/temp-f7355ed7/go-functions-example/site
Functions path:     /home/go/temp-f7355ed7/go-functions-example/functions
Configuration path: /home/go/temp-f7355ed7/go-functions-example/netlify.toml
Deploying to draft URL...
✔ Deploying functions from cache (use --skip-functions-cache to override)
✔ Finished hashing 2 files and 1 functions
✔ CDN requesting 0 files and 1 functions
⠼ (0/1) Uploading hello-lambda... ›   Warning: JSONHTTPError:  422
 ›   Warning:
{
  "name": "JSONHTTPError",
  "status": 422,
  "json": {
    "errors": "Invalid AWS Lambda parameters. Reserved keys used in this request"
  }
}

 ›   JSONHTTPError: Unprocessable Entity
```

(I also had to run a `go mod init` at the project's root otherwise it wouldn't build).

I don't think the issue has ever been building it. The issue is uploading it to AWS Lambda through the Netlify API.

Looking forward to the EOW update!

**Comment by eduardoboucas** [comment]
Hi again! This has been fixed in [version 8.11.0](https://github.com/netlify/cli/releases/tag/v8.11.0) of the Netlify CLI.

If you're still experiencing any problems, please let me know and I'll reopen the issue.

Thanks for bearing with me and apologies again for the frustrating experience here.
